[Zeppelin 3388] Refactor documentation for ZeppelinContext

sanjaydasgupta opened the pull request:
### What is this PR for?
This PR is a refactoring of the `ZepplinContext` documentation.

The description of `ZepplinContext` is now almost entirely contained within the _Spark_ interpreter's documentation ([spark.md](https://github.com/apache/zeppelin/blob/master/docs/interpreter/spark.md)). But `ZepplinContext` has many generic features that are available to all interpreters, and it is important for `ZepplinContext` to have a more visible and independent presence in the Zeppelin documentation.


### What type of PR is it?
[Documentation]

### Todos

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-3388

### How should this be tested?
Only documentation. 

### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No (This _is_ documentation)

zjffdu commented:
Thanks @sanjaydasgupta LGTM 
will merge it to branch-0.8 and master

felixcheung commented on docs/usage/other_features/zeppelin_context.md:
```diff
@@ -0,0 +1,230 @@
+In many programming-language based interpreters (e.g. Spark, Python, R) the zeppelin-context is available
+as a predefined variable `z` that can be used by directly invoking its methods.
+The methods available on the `z` object are described below.
+Interpreters based on programming languages like spark.dep, beam, etc. also provide the
```
we might need to clarify `beam` as `Apache Beam`. ditto for `Spark` perhaps as `Apache Spark`

sanjaydasgupta replied:
Done

felixcheung commented on docs/usage/other_features/zeppelin_context.md:
```diff
@@ -0,0 +1,230 @@
+
+## Usage in Programming Language Cells
+
+In many programming-language based interpreters (e.g. Spark, Python, R) the zeppelin-context is available
```
note, `Spark` is referred to as `Scala` below

sanjaydasgupta replied:
Clarified that Scala is used within the Spark cell

felixcheung commented on docs/usage/other_features/zeppelin_context.md:
```diff
@@ -0,0 +1,230 @@
+
+### Exploring Spark DataFrames
+The zeppelin-context provides a `show` method, which, using Zeppelin's `table` feature,
+can be used to nicely display a Spark DataFrame:
```
this was in the Spark interpreter page, should we add back the context here, "In the Spark interpreter, the zeppelin context"...

sanjaydasgupta replied:
Done. 
Also added a futuristic line saying `show()` will be progressively available in other interpreters.

felixcheung commented on docs/usage/other_features/zeppelin_context.md:
```diff
@@ -0,0 +1,230 @@
+Some interpreters use a subclass of `BaseZepplinContext` augmented with interpreter-specific
+functions. For example the dependency loader (%spark.dep) has `addRepo()`, `load()`, and other
+related methods that implement its functionality. Interpreter-specific functions are
+described within each interpreter's documentation.
```
maybe to clarify, `addRepo()` should be `z.addRepo()` etc.?

sanjaydasgupta replied:
Done

felixcheung commented on docs/usage/other_features/zeppelin_context.md:
```diff
@@ -0,0 +1,230 @@
+
+## Usage in Non-Programming Cells
+
+Cells with a non-programming interpreter (see table below) use a pattern match-and-replace 
```
I think the term "non-programming interpreter" is quite confusing.

felixcheung replied:
especially sql, jdbc, pig, python, are on the list.
or even cassandra, elasticsearch

sanjaydasgupta replied:
Used a different approach. Is this better?

felixcheung commented on docs/usage/other_features/zeppelin_context.md:
```diff
@@ -0,0 +1,230 @@
+
+|                   Non-Programming Interpreters                    |
+|-------------------------------------------------------------------|
+|spark.sql, bigquery, cassandra, elasticsearch, file, hbase, ignite, jdbc, kylin, livy, markdown, neo4j, pig, python, shell, zengine    |
```
listing this is quite error prone - this is too easy to forget to update. 
instead of the black list, how about we put the white list instead? ie. the ones that have object interpolation? is that shorter?

sanjaydasgupta replied:
This is in fact the white-list - but for 2 features: dynamic-forms and object-interpolation. 
The interpreters that implement object-interpolation now have a _(*)_ suffix which is explained in text below the table.

sanjaydasgupta commented:
Hi @felixcheung, have addressed your comments above.

Documentation written by a worldwide team of contributors can be notoriously ambiguous, and am happy to get a thorough review of this contribution.

Thanks for your time.

sanjaydasgupta commented:
This is how it looks when squeezed into a really narrow space:
![image](https://user-images.githubusercontent.com/477015/39469290-1a0e6f5e-4d55-11e8-984f-77814ee034b4.png)
And it resizes nicely with changing width too.

Thanks @felixcheung

zjffdu commented:
LGTM, merge if more comments
